src/tactic/lint.lean

```diff
@@ -361,6 +377,70 @@ meta def linter.has_inhabited_instance : linter :=
   errors_found := "TYPES ARE MISSING INHABITED INSTANCES",
   is_fast := ff }
 
+/-- Checks whether an instance can never be applied. -/
+meta def impossible_instance (d : declaration) : tactic (option string) := do
+  tt ← is_instance d.to_name | return none,
+  (binders, _) ← get_pi_binders_dep d.type,
+  let bad_arguments := binders.filter $ λ nb, nb.2.info ≠ binder_info.inst_implicit,
+  _ :: _ ← return bad_arguments | return none,
+  (λ s, some $ "Impossible to infer " ++ s) <$> print_arguments bad_arguments
+
+/-- A linter object for `impossible_instance`. -/
+@[linter, priority 1430] meta def linter.impossible_instance : linter :=
+{ test := impossible_instance,
+  no_errors_found := "All instances are applicable",
+  errors_found := "IMPOSSIBLE INSTANCES FOUND.\nThese instances have an argument that cannot be found during type-class resolution, and therefore can never succeed. Either mark the arguments with square brackets (if it is a class), or don't make it an instance" }
+
+/-- Checks whether the definition `nm` unfolds to a class. -/
+/- Note: Caching the result of `unfolds_to_class` by giving it an attribute (so that e.g. `vector_space` or `decidable_eq` would not be repeatedly unfold to check whether it is a class), did not speed up this tactic when executed on all of mathlib (and instead significantly slowed it down) -/
+meta def unfolds_to_class : name → tactic bool | nm :=
+if nm = `has_reflect then return tt else
+succeeds $ has_attribute `class nm <|> do
+  d ← get_decl nm,
+  tt ← unfolds_to_class d.value.lambda_body.pi_codomain.get_app_fn.const_name,
+  return 0
```

Comment:
It took me a minute to parse what's going on here -- `tactic bool` ending with `return 0` is confusing. Maybe using parens instead of `$` would make this easier to read?

Reply:
added a comment